File: 4. CNF Satisfiability (SAT).py
# Write a program using backtracking to solve the CNF-SAT problem (general Boolean satisfiability).
# This is the first proven NP-Complete problem.

from itertools import product

def evaluate_cnf(clauses, assignment):
    for clause in clauses:
        clause_value = False
        for literal in clause:
            var = abs(literal) - 1
            value = assignment[var]
            if literal < 0:
                value = not value
            clause_value |= value
        if not clause_value:
            return False
    return True

def solve_sat(n, clauses):
    for bits in product([False, True], repeat=n):
        if evaluate_cnf(clauses, bits):
            return True
    return False

# Example: (A ∨ ¬B) ∧ (B ∨ C)
clauses = [
    [1, -2],  # A ∨ ¬B
    [2, 3]    # B ∨ C
]
n = 3  # Variables: A, B, C

if solve_sat(n, clauses):
    print("Satisfiable ✅")
else:
    print("Unsatisfiable ❌")

"""
Time Complexity: O(2^n * m * k)
    - 2^n possible truth assignments
    - Each clause (m) has k literals to check

Space Complexity: O(n + m*k)
    - Store variable assignments and CNF structure
"""

File: 4. CNF Satisfiability (SAT).cpp
// Write a program using backtracking to solve the CNF-SAT problem (general Boolean satisfiability).
// This is the first proven NP-Complete problem.

#include <bits/stdc++.h>
using namespace std;

// Evaluate CNF clauses for a given variable assignment
bool evaluateCNF(const vector<vector<int>>& clauses, const vector<bool>& assignment)
{
    for (auto& clause : clauses)
    {
        bool clauseValue = false;
        for (int literal : clause)
        {
            int var = abs(literal) - 1;
            bool value = assignment[var];
            if (literal < 0)
                value = !value;  // handle negation
            clauseValue |= value;
        }
        if (!clauseValue)
            return false;  // any false clause => formula false
    }
    return true;
}

bool solveSAT(int n, const vector<vector<int>>& clauses)
{
    int total = 1 << n;  // 2^n combinations
    for (int mask = 0; mask < total; mask++)
    {
        vector<bool> assignment(n);
        for (int i = 0; i < n; i++) assignment[i] = (mask >> i) & 1;
        if (evaluateCNF(clauses, assignment))
            return true;
    }
    return false;
}

int main()
{
    // (A ∨ ¬B) ∧ (B ∨ C)
    vector<vector<int>> clauses = {
        {1, -2},  // A ∨ ¬B
        {2, 3}    // B ∨ C
    };

    int n = 3;  // Variables: A, B, C

    if (solveSAT(n, clauses))
        cout << "Satisfiable ✅" << endl;
    else
        cout << "Unsatisfiable ❌" << endl;

    return 0;
}

/*
Time Complexity: O(2^n * m * k)
    - 2^n assignments (n = number of variables)
    - m = number of clauses
    - k = average literals per clause

Space Complexity: O(n + m*k)
    - storing assignments and CNF clauses
*/
